File: 0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int maxrep = 0;
        int n = s.length();
        HashMap<Character, Integer> map = new HashMap<>();
        int i = 0, j = 0;

        while (i < n) {
            char ch = s.charAt(i);
            if (map.containsKey(ch)) {
                j= Math.max(j,map.get(ch)+1);
            }
            maxrep = Math.max(maxrep, i - j + 1);
            map.put(ch, i);
            i++;
        }
        return maxrep;
    }
}

File: 0003-longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<char> v;
        int maxLen = 0;

        for (char c : s) {
            auto it = find(v.begin(), v.end(), c);
            if (it != v.end()) {
                v.erase(v.begin(), it + 1);
            }
            v.push_back(c);
            maxLen = max(maxLen, (int)v.size());
        }

        return maxLen;
    }
};

